common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -0,0 +1,307 @@
+	// simplify things for ourselves by ignoring disabled watches and only
+	// actually adding them when ToggleDBusWatch() is called.
+	if (!dbus_watch_get_enabled(w)) {
+		return TRUE;
+	}
+
+	DBusClient *client = static_cast<DBusClient *>(data);
+	asio::posix::stream_descriptor *sd =
+		new asio::posix::stream_descriptor {DBusClient::GetAsioIoContext(client->loop_)};
+	boost::system::error_code ec;
+	sd->assign(dbus_watch_get_unix_fd(w), ec);
+	if (ec) {
+		log::Error("Failed to assign DBus FD to ASIO stream descriptor");
+		return FALSE;
+	}
+
+	unsigned int flags = dbus_watch_get_flags(w);
+	if (flags & DBUS_WATCH_READABLE) {
+		sd->async_wait(
+			asio::posix::stream_descriptor::wait_read,
+			[w, client, flags](boost::system::error_code ec) {
+				if (!dbus_watch_handle(w, flags)) {
+					log::Error("Failed to handle readable watch");
+				}
+				HandleDispatch(client->dbus_conn_.get(), DBUS_DISPATCH_DATA_REMAINS, client);
+			});
+	}
+	if (flags & DBUS_WATCH_WRITABLE) {
+		sd->async_wait(
+			asio::posix::stream_descriptor::wait_write,
+			[w, client, flags](boost::system::error_code ec) {
+				if (!dbus_watch_handle(w, flags)) {
+					log::Error("Failed to handle writable watch");
+				}
+				HandleDispatch(client->dbus_conn_.get(), DBUS_DISPATCH_DATA_REMAINS, client);
+			});
+	}
+	// Always watch for errors.
+	sd->async_wait(asio::posix::stream_descriptor::wait_error, [w](boost::system::error_code ec) {
+		if (!dbus_watch_handle(w, DBUS_WATCH_ERROR)) {
+			log::Error("Failed to handle error watch");
+		}
+	});
+
+	// Assign the stream_descriptor so that we have access to it in
+	// RemoveDBusWatch() and we can delete it.
+	dbus_watch_set_data(w, sd, NULL);
+	return TRUE;
+}
+
+static void RemoveDBusWatch(DBusWatch *w, void *data) {
+	asio::posix::stream_descriptor *sd =
+		static_cast<asio::posix::stream_descriptor *>(dbus_watch_get_data(w));
+	sd->cancel();
+	delete sd;
+	dbus_watch_set_data(w, NULL, NULL);
+}
+
+static void ToggleDBusWatch(DBusWatch *w, void *data) {
+	if (dbus_watch_get_enabled(w)) {
+		AddDBusWatch(w, data);
+	} else {
+		RemoveDBusWatch(w, data);
+	}
+}
+
+dbus_bool_t AddDBusTimeout(DBusTimeout *t, void *data) {
+	// See AddDBusWatch() for the details about this trick.
+	if (!dbus_timeout_get_enabled(t)) {
+		return TRUE;
+	}
+
+	DBusClient *client = static_cast<DBusClient *>(data);
+	asio::steady_timer *timer =
+		new asio::steady_timer {DBusClient::GetAsioIoContext(client->loop_)};
+	timer->expires_after(chrono::milliseconds {dbus_timeout_get_interval(t)});
+	timer->async_wait([t](boost::system::error_code ec) {
+		if (!dbus_timeout_handle(t)) {
+			log::Error("Failed to handle timeout");
+		}
+	});
+
+	dbus_timeout_set_data(t, timer, NULL);
+
+	return TRUE;
+}
+
+static void RemoveDBusTimeout(DBusTimeout *t, void *data) {
+	asio::steady_timer *timer = static_cast<asio::steady_timer *>(dbus_timeout_get_data(t));
+	if (timer != NULL) {
+		timer->cancel();
+		delete timer;
+	}
+	dbus_timeout_set_data(t, NULL, NULL);
```

Comment:
Should you call this first? Probably it doesn't matter in a single threaded context, but to me it seems more natural with the opposite order: Remove from list first, then delete, so that there is never an invalid object in there.